log, server: Create new log.EventLog() API

kyle-a-wong opened the pull request:
This PR introduces a new log.EventLog() API that gives
first class support for writing to the system.eventlog
table. This API lives in the log package and doesn't
have a dependency on the sql package, making it easy
to use across the repo.

This is implemented using a new set of interfaces in
the log package that allows the system and tenant
servers to register new event log writers. The logic
for writing to the event log table is the same as
that which is sql/event_log.go. For now the code is
duplicated, but it will be removed in a future commit
when this file is updated to use the new API.

Epic: [CRDB-42978](https://cockroachlabs.atlassian.net/browse/CRDB-42978)
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147769)
<!-- Reviewable:end -->

kyle-a-wong committed:
log, server: Create new log.EventLog() API

This PR introduces a new log.EventLog() API that gives
first class support for writing to the system.eventlog
table. This API lives in the log package and doesn't
have a dependency on the sql package, making it easy
to use across the repo.

This is implemented using a new set of interfaces in
the log package that allows the system and tenant
servers to register new event log writers. The logic
for writing to the event log table is the same as
that which is sql/event_log.go. For now the code is
duplicated, but it will be removed in a future commit
when this file is updated to use the new API.

Epic: CRDB-42978
Release note: None

kyle-a-wong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147769)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`pkg/util/log/eventlog/writer.go` line 112 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147769#-OS0vNs-2vdujVmpG7Ha:-OS0vNs-2vdujVmpG7Hb:b3sxovt) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a700477b95afc53c245a02f22dbd1e8f1c8f695/pkg/util/log/eventlog/writer.go#L112)):*
> ```Go
> }
> 
> func (e *Writer) asyncWrite(ctx context.Context, ev logpb.EventPayload) {
> ```

This code was lifted from sql/event_log.go with minor changes


<!-- Sent from Reviewable.io -->

angles-n-daemons reviewed (commented):
Couple of initial comments, looks like a great start though!

kyle-a-wong commented on pkg/server/server.go:
```diff
@@ -1835,6 +1836,21 @@ func (s *topLevelServer) PreStart(ctx context.Context) error {
+	// Creates and registers an event log sync for the system tenant. This will
+	// enable the ability to persist structured events to the system tenant's
+	// system.eventlog table.
+	eventTableSink := eventlog.NewWriter(s.node.execCfg.InternalDB, writeAsync, s.stopper, s.cfg.AmbientCtx, s.ClusterSettings())
```
Good idea. Done

dhartunian reviewed (approved):
> Reviewed 12 of 28 files at r1, 6 of 7 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147769)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @aa-joshi, @Abhinav1299, @alyshanjahani-crl, @angles-n-daemons, @arjunmahishi, and @kyle-a-wong)

___
*[`pkg/util/log/eventlog/writer.go` line 112 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147769#-OS0vNs-2vdujVmpG7Ha:-OTZKkSpAIdbIVGGTm6D:bm6bqv3) ([raw file](https://github.com/cockroachdb/cockroach/blob/6a700477b95afc53c245a02f22dbd1e8f1c8f695/pkg/util/log/eventlog/writer.go#L112)):*
<details><summary><i>Previously, kyle-a-wong (Kyle Wong) wrote…</i></summary><blockquote>

This code was lifted from sql/event_log.go with minor changes
</blockquote></details>

I added a TODO to unify the implementations.
___
*[`pkg/util/log/event_log.go` line 54 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147769#-OTXYR3O6SS2avc5g5xs:-OTXYR3O6SS2avc5g5xt:bnv09qh) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f37f5331249e9832a7f3bd110082416aecee5f2/pkg/util/log/event_log.go#L54)):*
> ```Go
> 	if eventLogWriterRegistry == nil {
> 		if buildutil.CrdbTestBuild {
> 			panic("OH NO")
> ```

nit: fix error message

(done)
___
*[`pkg/util/log/event_log.go` line 68 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147769#-OTXYSYy3Wf_D4pG1p4e:-OTXYSYzElkjPENFydSl:b-nz6in1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f37f5331249e9832a7f3bd110082416aecee5f2/pkg/util/log/event_log.go#L68)):*
> ```Go
> 	if eventLogWriterRegistry == nil {
> 		if buildutil.CrdbTestBuild {
> 			panic("OH NO")
> ```

nit: fix error message.

(done)
___
*[`pkg/util/log/event_log.go` line 184 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/147769#-OTTYVdI7AQ5-bkl87AW:-OTTYVdI7AQ5-bkl87AX:bxbx3v1) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f37f5331249e9832a7f3bd110082416aecee5f2/pkg/util/log/event_log.go#L184)):*
> ```Go
> }
> 
> func WithFatal() StructuredEventSettingsFunc {
> ```

we should probably not support `FATAL` for events as it's supposed to crash the node.

(I removed it)


<!-- Sent from Reviewable.io -->

dhartunian commented:
@angles-n-daemons can you take another look at this PR and let me know if you have further edits?

angles-n-daemons reviewed (approved):
nice, thanks both for this really good changeset.
